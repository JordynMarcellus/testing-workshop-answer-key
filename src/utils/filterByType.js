//TODO: past jord note -- don't pass in an empty array in the final code. this is a subtle bug that will show why it's good to test ðŸ¤“
export const filterByType = (arrayOfPokemon, type) =>
  arrayOfPokemon.reduce((filteredPokemonArray, pokemonAtArrayIndex) => {
    if (pokemonAtArrayIndex.types.includes(type)) {
      return [...filteredPokemonArray, pokemonAtArrayIndex];
    }
    return filteredPokemonArray;
  });
