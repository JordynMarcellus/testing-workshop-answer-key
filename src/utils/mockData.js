export const mockPikaData = {
  data: {
    name: "pikachu",
    types: [
      {
        type: {
          name: "electric",
        },
      },
    ],
    sprites: {
      front_default: "path/to/img",
    },
  },
};

export const mockGolemData = {
  data: {
    name: "golem",
    types: [
      {
        type: {
          name: "rock",
        },
      },
      {
        type: {
          name: "ground",
        },
      },
    ],
    sprites: {
      front_default: "path/to/img",
    },
  },
};
