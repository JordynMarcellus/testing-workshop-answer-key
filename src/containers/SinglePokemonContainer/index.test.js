import React from "react";
import renderer from "react-test-renderer";
import axios from "axios";
import { render, cleanup, waitForElement } from "react-testing-library";
import { SinglePokemonContainer, fetchPokemon } from "./";
import { mockPikaData } from "../../utils/mockData";
import "jest-dom/extend-expect";

jest.mock("axios");

describe("SinglePokemonContainer", () => {
  it("successfully calls the async data fetching function", async () => {
    axios.mockResolvedValue(mockPikaData);
    const loadingChange = jest.fn();
    const pokemonChange = jest.fn();
    const expectedUrl = "https://pokeapi.co/api/v2/pokemon/pikachu";
    expect.assertions(3);
    await fetchPokemon("pikachu", loadingChange, pokemonChange);
    await expect(axios).toBeCalledWith(expectedUrl);
    expect(loadingChange).toBeCalledWith(false);
    expect(pokemonChange).toBeCalledWith(mockPikaData.data);
  });
  it("passes an error condition", async () => {
    axios.mockResolvedValue(mockPikaData);
    const loadingChange = jest.fn();
    const pokemonChange = jest.fn();
    const expectedUrl = "https://pokeapi.co/api/v2/pokemon/pikachu";
    expect.assertions(3);
    await fetchPokemon("pikachu", loadingChange, pokemonChange);
    await expect(axios).toBeCalledWith(expectedUrl);
    expect(loadingChange).toBeCalledWith(false);
    expect(pokemonChange).toBeCalledWith(mockPikaData.data);
  });
  describe("useEffect integration tests", () => {
    afterEach(cleanup);
    it("renders the loading state", () => {
      const originalError = console.error;
      console.error = jest.fn();

      const { container } = render(
        <SinglePokemonContainer pokemonName={"Pikachu"} />
      );
      expect(container.firstChild).toMatchSnapshot();
      expect(container).toHaveTextContent("Loading");
      console.error = originalError;
    });
    it("renders the resolved state", async () => {
      const originalError = console.error;
      console.error = jest.fn();

      const { getByText, container } = render(
        <SinglePokemonContainer pokemonName={"Pikachu"} />
      );
      axios.mockResolvedValue(mockPikaData);
      await waitForElement(() => getByText("pikachu"));
      expect.assertions(2);
      expect(container.firstChild).toMatchSnapshot();
      expect(container).toHaveTextContent("pikachu");

      console.error = originalError;
    });
  });
});
