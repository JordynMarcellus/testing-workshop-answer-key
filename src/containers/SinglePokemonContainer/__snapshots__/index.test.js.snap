// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SinglePokemonContainer useEffect integration tests renders the loading state 1`] = `
<div>
  Loading
</div>
`;

exports[`SinglePokemonContainer useEffect integration tests renders the resolved state 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  max-width: 300px;
}

<article
  class="emotion-0 emotion-1"
>
  <h1>
    pikachu
  </h1>
  <ul>
    <span>
      electric
    </span>
  </ul>
  <img
    alt="My chonky fave pikachu"
    src="path/to/img"
    title="My chonky fave pikachu"
  />
</article>
`;
